### ISSUE TITLE
thousands of "awk" and "sed" zombie processes with the native procstat plugin on MacOS

### ISSUE THREAD

**Issue by jimis** [influxdata/telegraf#6265]
### Relevant telegraf.conf:

```toml
[[inputs.procstat]]
  exe = "coin-"
  pid_finder = "native"
```

### System info:

Telegraf 1.11.4
MacOS X 10.13

### More details

"sed" and "awk" processes flood the process table until the system comes down to a halt. I'm quite sure it's because of the procstat plugin, as the problem went away once I used "pgrep" as the pid_finder.

**Comment by jimis** [comment]
AFAICT the procstat plugin, using the "native" pid_finder, it runs `lsof` and then parses it with `sed` and `awk` (and not reaping them correctly apparently). I was expecting the plugin wouldn't even fork, since it's called "native" !

**Comment by danielnelson** [comment]
> I was expecting the plugin wouldn't even fork, since it's called "native" !

Yeah, me too.

I looked into this the native pid_finder is using gopsutil for cross platform support and calls [process.Process()](https://github.com/shirou/gopsutil/blob/master/process/process_darwin.go#L576) and [process.Exe()](https://github.com/shirou/gopsutil/blob/master/process/process_darwin.go#L576).

On Linux this doesn't exec any system utilities, but on Darwin it does appear to use ps and lsof, though I'm not seeing awk/sed called anywhere.

It would be really helpful if someone could work with the github.com/shirou/gopsutil team to improve these operations on Darwin.

**Comment by jimis** [comment]
Apparently it used to call awk and sed: https://github.com/shirou/gopsutil/pull/684

Now it only calls `lsof`, which is quite slow process in my experience, but still it is better than what it used to do. Maybe an update of gopsutil is all it takes to fix the issue?

**Comment by danielnelson** [comment]
In the upcoming Telegraf 1.12 (or the [nightly builds](https://github.com/influxdata/telegraf#nightly-builds)) we have a newer gopsutil.

I agree it would be nice to have an even more performant solution that doesn't depend on any external tools, so I still encourage anyone with interest and expertise in this area to open a pull request with gopsutil.

**Comment by jimis** [comment]
Thanks, I'm closing this as it seems resolved in 1.12: it no longer calls `awk` and `sed`, but only `lsof`, and it leaves no zombies behind.
